cli/internal/run/run.go

```diff
@@ -311,161 +274,84 @@ func (r *run) runOperation(ctx gocontext.Context, g *graph.CompleteGraph, rs *ru
 		}
 	}
 
+	// Graph Run
 	if rs.Opts.runOpts.graphFile != "" || rs.Opts.runOpts.graphDot {
-		graph := engine.TaskGraph
-		if r.opts.runOpts.singlePackage {
-			graph = filterSinglePackageGraphForDisplay(engine.TaskGraph)
-		}
-		visualizer := graphvisualizer.New(r.base.RepoRoot, r.base.UI, graph)
-
-		if rs.Opts.runOpts.graphDot {
-			visualizer.RenderDotGraph()
-		} else {
-			err := visualizer.GenerateGraphFile(rs.Opts.runOpts.graphFile)
-			if err != nil {
-				return err
-			}
-		}
-	} else if rs.Opts.runOpts.dryRun {
-		tasksRun, err := r.executeDryRun(ctx, engine, g, tracker, rs)
-		if err != nil {
-			return err
-		}
-		packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
-		sort.Strings(packagesInScope)
-		if rs.Opts.runOpts.dryRunJSON {
-			var rendered string
-			if r.opts.runOpts.singlePackage {
-				rendered, err = renderDryRunSinglePackageJSON(tasksRun)
-			} else {
-				rendered, err = renderDryRunFullJSON(tasksRun, packagesInScope)
-			}
-			if err != nil {
-				return err
-			}
-			r.base.UI.Output(rendered)
-		} else {
-			if err := displayDryTextRun(r.base.UI, tasksRun, packagesInScope, g.PackageInfos, r.opts.runOpts.singlePackage); err != nil {
-				return err
-			}
-		}
-	} else {
-		packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
-		sort.Strings(packagesInScope)
-		if r.opts.runOpts.singlePackage {
-			r.base.UI.Output(fmt.Sprintf("%s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", ")))))
-		} else {
-			r.base.UI.Output(fmt.Sprintf(ui.Dim("• Packages in scope: %v"), strings.Join(packagesInScope, ", ")))
-			r.base.UI.Output(fmt.Sprintf("%s %s %s", ui.Dim("• Running"), ui.Dim(ui.Bold(strings.Join(rs.Targets, ", "))), ui.Dim(fmt.Sprintf("in %v packages", rs.FilteredPkgs.Len()))))
-		}
-		return r.executeTasks(ctx, g, rs, engine, packageManager, tracker, startAt)
+		return GraphRun(ctx, rs, engine, r.base)
 	}
-	return nil
-}
 
-func renderDryRunSinglePackageJSON(tasksRun []hashedTask) (string, error) {
-	singlePackageTasks := make([]hashedSinglePackageTask, len(tasksRun))
-	for i, ht := range tasksRun {
-		singlePackageTasks[i] = ht.toSinglePackageTask()
-	}
-	dryRun := &struct {
-		Tasks []hashedSinglePackageTask `json:"tasks"`
-	}{singlePackageTasks}
-	bytes, err := json.MarshalIndent(dryRun, "", "  ")
-	if err != nil {
-		return "", errors.Wrap(err, "failed to render JSON")
-	}
-	return string(bytes), nil
-}
+	packagesInScope := rs.FilteredPkgs.UnsafeListOfStrings()
+	sort.Strings(packagesInScope)
+	// Initiate analytics and cache
+	analyticsClient := r.initAnalyticsClient(ctx)
+	defer analyticsClient.CloseWithTimeout(50 * time.Millisecond)
+	turboCache, err := r.initCache(ctx, rs, analyticsClient)
```

Comment:
Whatever receives the turboCache must now make sure it's shutdown

Reply:
Yeah, that's a bit weird. DryRun and RealRun both receive it. They were both individually calling `initCache()` before and I felt like it was not necessary to duplicate, but also, it really helped boil down the different between `DryRun` and `RealRun` to the bare minimum. Only reason I couldn't call `defer turbocache.shutdown()` here is because it happens in different ways for each (and I'm not sure why)

Reply:
Oops, this was meant to be a comment just for myself. It does seem like the best we can do until the shutdown logic gets unified.